### ISSUE TITLE
Beautify Bazel configuration errors

### ISSUE THREAD

**Issue by gregestren** [bazelbuild/bazel#11984]
**tl;dr:** apply [beautiful error](https://bazel.build/designs/2016/05/23/beautiful-error-messages.html)  principles to build failures from analysis, `select()`, configurations, user-defined flags, transitions, and so on. 

---
We have enough experience to identify where current errors fall short. There are straightforward ways we make them clearer, more actionable, and more closely connected to relevant documentation.

I'm starting this for configuration-related errors since this is the Bazel area I know best. More broadly I'd like to extend these principles to Bazel error culture as a whole in the interest of making Bazel a more accessible build tool.

Crowd-sourced input on confusing messages welcome!

**Comment by gregestren** [comment]
@jin

**Comment by mobileink** [comment]
If you're collecting error messages that need fixing, here's another one:

 ```
ERROR: <root>/src/BUILD.bazel:596:13: output 'src/refList.o' was not created
ERROR: <root>/src/BUILD.bazel:596:13: not all outputs were created or valid
```

This is not much better than "Your build failed".  It will be totally opaque to users who do not know that "output" has special meaning for Bazel rules.  It does not mean the compile failed; it might have failed, but it might have succeeded yet failed to put the output in the right place.  Worst of all is "[not] created or valid" - doesn't Bazel know which?

Best,

Gregg

**Comment by gregestren** [comment]
Thanks, Gregg. Excellent and subtle point on the "A or B" phrasing. I'll dig in when I get a chance to see what options we have for those errors.

**Comment by mobileink** [comment]
Here's another useless error message:

```
ERROR: Failed to load Starlark extension '@bazel_skylib//lib:paths.bzl'.
Cycle in the workspace file detected. This indicates that a repository is used prior to being defined.
The following chain of repository dependencies lead to the missing definition.
 - @bazel_skylib
This could either mean you have to add the '@bazel_skylib' repository with a statement like `http_archive` in your WORKSPACE file (note that transitive dependencies are not added automatically), or move an existing definition earlier in your WORKSPACE file.
ERROR: cycles detected during target parsing
```

I have `@bazel_skylib//lib:paths.bzl` in a dozen files; which is it?  Surely Bazel knows.  I also have multiple workspaces. Bazel should be able to tell me where the problem is.

Gregg

**Comment by gregestren** [comment]
@mobileink - I just submitted https://github.com/bazelbuild/bazel/commit/3d33d7b2dea8b4ea92162cc18b1ed2c7c31ad6e6 and looking for another error to polish. 

For https://github.com/bazelbuild/bazel/issues/11984#issuecomment-703849350, could your version of the ideal message? In https://github.com/bazelbuild/bazel/commit/3d33d7b2dea8b4ea92162cc18b1ed2c7c31ad6e6 's description I laid out principles I'm trying to introduce: basically make error messages look like great StackOverflow answers (clear, helpful, readable, concise). 

Re: implementation, I'm not sure how much of the relevant context the code at that stage has. That error is generated at https://github.com/bazelbuild/bazel/blob/a790c172fbc15bafc20417a6e6a235f4abd9fc9a/src/main/java/com/google/devtools/build/lib/skyframe/SkyframeActionExecutor.java#L1599 and https://github.com/bazelbuild/bazel/blob/a790c172fbc15bafc20417a6e6a235f4abd9fc9a/src/main/java/com/google/devtools/build/lib/skyframe/SkyframeActionExecutor.java#L1194

**Comment by catskul** [comment]
Here's one that's not confusing exactly, but just hard to read and easy to misread.
```
ERROR: /home/catskul/repos/av/red/RedInfManager/tests/BUILD:83:8: in cc_test rule //red/RedInfManager/tests:MTRConfigTest: target '//red/Tasks/Inf:RedInfTaskConfig' is not visible
from target '//red/RedInfManager/tests:MTRConfigTest'. Check the visibility declaration of the former target if you think the dependency is legitimate
```

Would be much more readable as:
```
ERROR: 
    /home/catskul/repos/av/red/RedInfManager/tests/BUILD:83:8: 
        in cc_test rule //red/RedInfManager/tests:MTRConfigTest: 
                target '//red/Tasks/Inf:RedInfTaskConfig' 
            is not visible from 
                target '//red/RedInfManager/tests:MTRConfigTest'. 
    Check the visibility declaration of the former target if you think the dependency is legitimate
```

**Comment by gregestren** [comment]
This is great. Thanks for the suggestion @catskul.

Some basic formatting standards could go a long way.

**Comment by bcsgh** [comment]
@gregestren regarding the "was not created" error message, what would improve things the most for me would be some help for cases where the action, worked and the expected file was generated, just with the wrong name or in the wrong location. I've on more occasions than I care to count stuffed `postfix_script = "find . >&2 && false",` into an external build rule just to try to figure out what it actually did.

Just a working directory listing (or a listing of new files that the action created) would be an improvement. If that's too verbose, then dump that to a file and tell me where it is. For bonus points; pick the 3-10 closes "matches" and ask "did you want...".

> "Nothing is a failure if you learn something you then use use to make the next attempt better."

In this case, while I learn something from the default message, there is nothing I can do with it to make the next attempt better.

**Comment by bcsgh** [comment]
I'm thinking of the cast where bazel has no clue what went wrong, only that no result ended up in the expected location.

I'm not expecting bazel to log file access or anything but just give me some actionable information about the final state. Something as simple as `find` from the rule's cwd would be nice. Even just dumping a path where I can do that from on my own would be an improvement. (Does bazel delete those directories on failure? I don't ever remember being able to locate them after the fact unless things worked.)

**Comment by gregestren** [comment]
You're doing local builds (no remote executor)?

Have you tried `$ bazel build //foo -s`?

**Comment by bcsgh** [comment]
I've tried `--subcommands` a few times, but IIRC I was never able to find the working directory from failed build steps listed in there. Even if it is, it's too verbose to be a practical solution to the problem here.

While I'd like to see at a minimum is the existing error message:
```
output 'src/refList.o' was not created
```
Becomes something like 
```
output 'src/refList.o' was not created in directory /some/path/to/steps/working/dir
```
Where  `find /some/path/to/steps/working/dir -type f` has a good chance of giving me something useful.
